File: scripts/fetch-quran.js
/**
 * Simple script to fetch full Quran from Al-Quran Cloud API
 * No API key required - completely free
 * 
 * Run: node scripts/fetch-quran.js
 */

const fs = require('fs');
const path = require('path');
const https = require('https');

const DATA_DIR = path.join(__dirname, '..', 'data');

// Ensure data directory exists
if (!fs.existsSync(DATA_DIR)) {
    fs.mkdirSync(DATA_DIR, { recursive: true });
}

function httpGet(url) {
    return new Promise((resolve, reject) => {
        https.get(url, (res) => {
            let data = '';
            res.on('data', (chunk) => { data += chunk; });
            res.on('end', () => {
                if (res.statusCode === 200) {
                    resolve(JSON.parse(data));
                } else {
                    reject(new Error(`HTTP ${res.statusCode}`));
                }
            });
        }).on('error', reject);
    });
}

async function downloadQuran() {
    console.log('Downloading full Quran from Al-Quran Cloud API...\n');
    const passages = [];
    
    // Al-Quran Cloud API - free, no API key needed
    const baseUrl = 'https://api.alquran.cloud/v1';
    
    // Get all 114 surahs
    for (let surahNum = 1; surahNum <= 114; surahNum++) {
        try {
            // Using Pickthall translation (public domain, 1930)
            const url = `${baseUrl}/surah/${surahNum}/en.pickthall`;
            process.stdout.write(`\rFetching Surah ${surahNum}/114...`);
            
            const response = await httpGet(url);
            
            if (response.code === 200 && response.data) {
                const surah = response.data;
                const surahName = surah.englishName || surah.name;
                
                // Process each ayah (verse)
                surah.ayahs.forEach((ayah) => {
                    // Clean HTML tags if any
                    const text = ayah.text.replace(/<[^>]*>/g, '').trim();
                    
                    passages.push({
                        reference: `${surahName} ${surahNum}:${ayah.numberInSurah}`,
                        text: text,
                        context: `Surah ${surahName} (Chapter ${surahNum})`
                    });
                });
            }
            
            // Small delay to avoid rate limiting
            await new Promise(resolve => setTimeout(resolve, 200));
            
        } catch (error) {
            console.error(`\nError fetching Surah ${surahNum}:`, error.message);
        }
    }
    
    console.log(`\n\nDownloaded ${passages.length} Quran verses`);
    return passages;
}

async function main() {
    try {
        const passages = await downloadQuran();
        
        if (passages.length > 0) {
            const filePath = path.join(DATA_DIR, 'quran.json');
            const data = { passages };
            fs.writeFileSync(filePath, JSON.stringify(data, null, 2));
            console.log(`\n✅ Saved ${passages.length} passages to ${filePath}`);
            console.log(`\nFile size: ${(fs.statSync(filePath).size / 1024 / 1024).toFixed(2)} MB`);
        } else {
            console.error('No passages downloaded');
        }
    } catch (error) {
        console.error('Error:', error);
    }
}

if (require.main === module) {
    main();
}

module.exports = { downloadQuran };

File: scripts/fetch-quran.py
#!/usr/bin/env python3
"""
Download full Quran from Al-Quran Cloud API (free, no API key needed)
Uses Pickthall translation (public domain, 1930)

Run: python3 scripts/fetch-quran.py
"""

import json
import os
import time
import urllib.request
import urllib.error
import ssl

DATA_DIR = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'data')

# Ensure data directory exists
os.makedirs(DATA_DIR, exist_ok=True)

def fetch_url(url):
    """Fetch data from URL"""
    try:
        # Create unverified context for SSL (sandbox environments)
        ssl_context = ssl.create_default_context()
        ssl_context.check_hostname = False
        ssl_context.verify_mode = ssl.CERT_NONE
        
        with urllib.request.urlopen(url, context=ssl_context) as response:
            return json.loads(response.read().decode('utf-8'))
    except urllib.error.HTTPError as e:
        print(f"HTTP Error {e.code}: {url}")
        return None
    except Exception as e:
        print(f"Error fetching {url}: {e}")
        return None

def download_quran():
    """Download all 114 surahs of the Quran"""
    print("Downloading full Quran from Al-Quran Cloud API...\n")
    passages = []
    
    base_url = "https://api.alquran.cloud/v1"
    
    # Get all 114 surahs
    for surah_num in range(1, 115):
        try:
            # Using Pickthall translation (public domain, 1930)
            url = f"{base_url}/surah/{surah_num}/en.pickthall"
            print(f"\rFetching Surah {surah_num}/114...", end='', flush=True)
            
            response = fetch_url(url)
            
            if response and response.get('code') == 200 and response.get('data'):
                surah = response['data']
                surah_name = surah.get('englishName') or surah.get('name', f'Surah {surah_num}')
                
                # Process each ayah (verse)
                for ayah in surah.get('ayahs', []):
                    # Clean HTML tags if any
                    text = ayah.get('text', '').replace('<i>', '').replace('</i>', '').strip()
                    text = ' '.join(text.split())  # Normalize whitespace
                    
                    passages.append({
                        "reference": f"{surah_name} {surah_num}:{ayah.get('numberInSurah', '')}",
                        "text": text,
                        "context": f"Surah {surah_name} (Chapter {surah_num})"
                    })
            
            # Small delay to avoid rate limiting
            time.sleep(0.2)
            
        except Exception as e:
            print(f"\nError fetching Surah {surah_num}: {e}")
    
    print(f"\n\nDownloaded {len(passages)} Quran verses")
    return passages

def main():
    try:
        passages = download_quran()
        
        if passages:
            file_path = os.path.join(DATA_DIR, 'quran.json')
            data = {"passages": passages}
            
            with open(file_path, 'w', encoding='utf-8') as f:
                json.dump(data, f, indent=2, ensure_ascii=False)
            
            file_size = os.path.getsize(file_path) / 1024 / 1024
            print(f"\n✅ Saved {len(passages)} passages to {file_path}")
            print(f"File size: {file_size:.2f} MB")
        else:
            print("No passages downloaded")
            
    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    main()
